go/vumitools/middleware.py

```diff
@@ -603,6 +604,151 @@ def handle_failure(self, failure, connector_name):
+    :param str op_mode:
+        What mode to operate in, options are `passive` or `active`.
+        Defaults to passive.
+        *passive*:  assumes the middleware connector names are to be used as
+                    the names for metrics publishing.
+        *active*:   assumes that the individual messages are to be inspected
+                    for their `transport_name` values.
+
+        NOTE:   This does not apply for events or failures, the connectors
+                names are always used for those since those message types are
+                not guaranteed to have a `transport_name` value.
+    :param list metric_connectors:
+        List of connector names to fire metrics for. Useful for when wrapping
+        dispatchers with many connectors, only a subset of which should
+        generate metrics. Defaults to all connectors.
+    """
+
+    CONFIG_CLASS = ConversationMetricsMiddlewareConfig
+
+    KNOWN_MODES = frozenset(['active', 'passive'])
+
+    def validate_config(self):
+        self.manager_name = get_go_metrics_prefix()+"campaigns"
+        self.op_mode = self.config.op_mode
+        if self.op_mode not in self.KNOWN_MODES:
+            raise ConfigError('Unknown op_mode: %s' % (
+                self.op_mode,))
+        self.metric_connectors_specified = (
+            self.config.metric_connectors is not None)
+        self.metric_connectors = set(self.config.metric_connectors or [])
+
+    @inlineCallbacks
+    def setup_middleware(self):
+        self.validate_config()
+        self.metric_publisher = yield self.worker.start_publisher(
+            MetricPublisher)
+        # We don't use a VumiApi here because we don't have a Riak config for
+        # it.
+        self.redis = yield TxRedisManager.from_config(
+            self.config.redis_manager)
+        self.metric_manager = MetricManager(
+            self.manager_name + '.', publisher=self.metric_publisher)
+        self.metric_manager.start_polling()
+
+    def teardown_middleware(self):
+        self.metric_manager.stop_polling()
+        return self.redis.close_manager()
+
+    def get_or_create_metric(self, name, metric_class, *args, **kwargs):
+        """
+        Get the metric for `name`, create it with
+        `metric_class(*args, **kwargs)` if it doesn't exist yet.
+        """
+        if name not in self.metric_manager:
+            self.metric_manager.register(metric_class(name, *args, **kwargs))
+        return self.metric_manager[name]
+
+    def get_counter_metric(self, name):
+        return self.get_or_create_metric(name, Count)
+
+    def increment_counter(self, prefix, message_type):
+        metric = self.get_counter_metric('%s.%s' % (prefix, message_type))
+        metric.inc()
+
+    def get_acc_key(self, msg):
+        mdh = MessageMetadataHelper(
+            self.vumi_api, msg, conversation_cache=self._conversation_cache)
+        acc_key = yield mdh.get_account_key()
+        returnValue(acc_key)
```

Comment:
Functions that use `yield` to yield on deferreds need the `@inlineCallbaks` decorator.